DataFormats/PatCandidates/src/PackedCandidate.cc

```diff
@@ -375,3 +375,38 @@ void pat::PackedCandidate::setHcalFraction(float p) {
 void pat::PackedCandidate::setIsIsolatedChargedHadron(bool p) {
   isIsolatedChargedHadron_ = p;
 }
+
+void pat::PackedCandidate::setDTimeAssociatedPV(float aTime, float aTimeError) {
+    if (aTime == 0 && aTimeError == 0) {
+        packedTime_ = 0; packedTimeError_ = 0;
+    } else if (aTimeError == 0) {
+        packedTimeError_ = 0;
+        packedTime_ = packTimeNoError(aTime);
+    } else {
+        packedTimeError_ = packTimeError(aTimeError);
+        aTimeError = unpackTimeError(packedTimeError_); // for reproducibility
+        packedTime_ = packTimeWithError(aTime, aTimeError);
+    }
+}
+
+/// static to allow unit testing
+uint8_t pat::PackedCandidate::packTimeError(float timeError) { return timeError <= 0 ? 0 : std::max<uint8_t>( std::min(std::round(std::ldexp(std::log2(timeError/0.002f), +5)), 255.f), 1); }
+float pat::PackedCandidate::unpackTimeError(uint8_t timeError) { return timeError > 0 ? 0.002f * std::exp2(std::ldexp(float(timeError),-5)) : -1.0f; }
+float pat::PackedCandidate::unpackTimeNoError(int16_t time) { return (time > 0 ? 0.0002f : -0.0002f) * std::exp2(std::ldexp(float(std::abs(time)),-6)); }
+int16_t pat::PackedCandidate::packTimeNoError(float time) { return (time > 0 ? +1 : -1)*std::min(std::round(std::ldexp(std::log2(std::abs(time/0.0002f)),+6)), 32767.f); }
```

Comment:
better change magic numbers to constexpr

what's the rationale in this packing?
I see that small numbers below the magic 0.0002f are not handled.
take an example time_original = 0.0002f/1024: time_packed = -640.
then unpackTimeNoError(-64) gives -0.24.

Can this be just a float32to16 from libminifloat?

Reply:
@gpetruc can you answer Slava's question about the packing?